[WebContent] Fix Integer Underflow in xmlReadMemory

zakariaridouh opened the pull request:
#### 66bfd48c94bf04db92322e41ea8b78d64b1c775e
<pre>
[WebContent] Fix Integer Underflow in xmlReadMemory
<a href="https://bugs.webkit.org/show_bug.cgi?id=304140">https://bugs.webkit.org/show_bug.cgi?id=304140</a>
<a href="https://rdar.apple.com/161583041">rdar://161583041</a>

Reviewed by Brandon Stewart and Ryosuke Niwa.

Fix some potential int underflows before passing the values to `libxml2`
by adding explicit bounds checks using INT_MAX in three locations:
- `XSLTProcessorLibxslt.cpp`: XSLT document loading
- `XMLDocumentParserLibxml2.cpp`: XML document parsing
- `InitDataRegistry.cpp`: PlayReady initialization data parsing (GStreamer)

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData):
* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlDocPtrForString):

Canonical link: <a href="https://commits.webkit.org/305255@main">https://commits.webkit.org/305255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/222663f62a5213a7570c67c6bb806ce7fdf0379e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136728 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9088 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144459 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89701 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9374f15d-47b2-485f-b85e-32757cfeb302) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138600 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104557 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76348 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Failed to checkout and rebase branch from PR 55372") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d552fb09-8946-42e6-9eff-22db8e6c4d55) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85396 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6801 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4490 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5048 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116121 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147216 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8771 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41271 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112911 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8789 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113241 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29007 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6722 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118801 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62923 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8819 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36855 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72385 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8759 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8611 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
```diff
@@ -204,7 +204,9 @@ bool isPlayReadySanitizedInitializationData(const SharedBuffer& buffer)
 
     auto xmlPayload = buffer.span().subspan(startTag);
-    xmlDocPtr protectionDataXML = xmlReadMemory(reinterpret_cast<const char*>(xmlPayload.data()), xmlPayload.size_bytes(), "protectionData", "utf-16", 0);
+    if (xmlPayload.size_bytes() > INT_MAX)
```
C++ does have `std::numeric_limits<int>::max()`, which we might want to consider instead.

stwrt reviewed (commented):
Also, based on the title checking for > max() would be an overflow right?

zakariaridouh committed:
[WebContent] Fix Integer Underflow in xmlReadMemory
https://bugs.webkit.org/show_bug.cgi?id=304140
rdar://161583041

Reviewed by Brandon Stewart and Ryosuke Niwa.

Fix some potential int underflows before passing the values to `libxml2`
by adding explicit bounds checks using INT_MAX in three locations:
- `XSLTProcessorLibxslt.cpp`: XSLT document loading
- `XMLDocumentParserLibxml2.cpp`: XML document parsing
- `InitDataRegistry.cpp`: PlayReady initialization data parsing (GStreamer)

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData):
* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlDocPtrForString):

Canonical link: https://commits.webkit.org/305255@main
